### PR TITLE
Implement packet coalescing in the local transport

### PR DESCRIPTION
**Pull request by chrisvest**
Motivation:
It is much more efficient to process data in batches, and this is true as well for data passing through the local transport, i.e. LocalChannel. Applications often write many small buffers to their channels, but expect to process fewer, larger buffers when receiving data from the network. In fact, when the SslHandler is in the pipeline, we can observe a 100x difference in performance when coalescing packets into bigger buffers before processing, compared to processing packets individually. The real transports, like NIO and the native transports, all operate like this.

Modification:
The LocalChannel will now inspect incoming messages, and opportunistically coalesce small Buffers into bigger ones before sending them down the pipeline.

Result:
The local transport now sees improved batch processing behaviour. For instance, the Http2ThroughputBenchmark is made a 100 times faster when using SSL over the local transport.

### CODE REVIEW

**Comment by normanmaurer** [comment]
That's pretty cool... Want to also do something similar for 4.1 ?

**Comment by franz1981** [comment]
When I read this I though about a different thing, but likely similar as a concept: see https://github.com/eclipse-vertx/vertx-grpc/pull/68

TLDR 
3 efficiency problems related coalescing:
- composite heap buffers benefit from being copied into a single direct buffer, avoiding write syscalls, but have very bad ping-pong of types/bound checks to perform the copy.
- direct composite ones won't benefit of this, and use writev, which is still bad depending how small are the components
- users usually perform many small writers and flush; currently we do not optimize to coalesce them in a single buffer (this is similar to the problem explained here)

I believe a mix of smart coalescing in the transport and. At the composite buffer api would help and improve performance; we can turn this into a discussion or a call if you want...
I am currently on PTO till 4 sept, anywah
